### ISSUE TITLE
Train obb model difficult to convergr

### ISSUE THREAD

**Issue by colchicinewf** [ultralytics/ultralytics#24920]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When training an OBB model on my own dataset with version 8.4.47, I found it difficult to converge. The first few epochs were normal, but later the loss became NaN. However, when I trained with version 8.4.12 previously, it worked fine. So, I compared the training performance of versions 8.4.12 and 8.4.60 (the latest version at the time of comparison) on DOTA v1.0 and found that the training results of version 8.4.60 indeed seemed problematic. Here are my experimental steps:  
1. First, I used the following code to crop the DOTA v1.0 data. (Correspondingly, I modified the path in DOTAv1.yaml to DOTAv1.0-split.)  
```
from ultralytics.data.split_dota import split_test, split_trainval

# Split train and val set, with labels.
split_trainval(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
)
# Split test set, without labels.
split_test(
    data_root="path/to/DOTAv1.0/",
    save_dir="path/to/DOTAv1.0-split/",
    rates=[0.5, 1.0, 1.5],  # multiscale
    gap=500,
```

3. I trained using the same code.  
4. Below are the training results for the two versions. For version 8.4.12, the best model achieved mAP50: 0.745 and mAP50-95: 0.597, which appears normal. However, for version 8.4.60, the best model only achieved mAP50: 0.548 and mAP50-95: 0.405, and the last model only reached 0.0943 and 0.080.

model  | mAP50 | mAP50-95
-- | -- | --
8.4.12 best | 0.745 | 0.597
 8.4.60 best | 0.548 | 0.405
 8.4.60 last | 0.0943 |  0.080

V8.4.12
<img width="1200" height="514" alt="Image" src="https://github.com/user-attachments/assets/7916028c-6c0a-4b5b-832e-a0d2bf0407d4" />
V8.4.60
<img width="2800" height="1200" alt="Image" src="https://github.com/user-attachments/assets/f49d1ff5-a687-49f1-ad2e-ba05aa57a325" />

### Environment

Ultralytics 8.4.60 🚀 Python-3.10.14 torch-2.5.1+cu118 CUDA:0 (NVIDIA A40, 45434MiB)
Setup complete ✅ (255 CPUs, 503.2 GB RAM, 0.0/10.0 GB disk)

OS                     Linux-5.4.132-2.pclimit.el7.x86_64-x86_64-with-glibc2.35
Environment            Docker
Python                 3.10.14
Install                git
Path                   /netds2/groupdata/medcvocralgorithm/10_Private/wang_fei/00_work/ultralytics-v8.4.60/ultralytics
RAM                    503.22 GB
Disk                   0.0/10.0 GB
CPU                    AMD EPYC 7713 64-Core Processor
CPU count              255
GPU                    NVIDIA A40, 45434MiB
GPU count              1
CUDA                   11.8

numpy                  ✅ 1.26.4>=1.23.0
matplotlib             ✅ 3.9.2>=3.3.0
opencv-python          ✅ 4.10.0.84>=4.6.0
pillow                 ✅ 11.0.0>=7.1.2
pyyaml                 ✅ 6.0.1>=5.3.1
requests               ✅ 2.32.3>=2.23.0
scipy                  ✅ 1.14.1>=1.4.1
torch                  ✅ 2.5.1+cu118>=1.8.0
torch                  ✅ 2.5.1+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.20.1+cu118>=0.9.0
psutil                 ✅ 5.9.8>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.19>=2.0.18

### Minimal Reproducible Example
```
import os
# os.environ['TORCH_DISTRIBUTED_DEBUG'] = 'DETAIL'
os.environ['NCCL_BLOCKING_WAIT'] = '0'  
os.environ['YOLO_CONFIG_DIR'] = os.path.dirname(os.path.abspath(__file__))
os.environ['YOLO_OFFLINE'] = 'true' 

os.system('yolo checks')

import argparse
import ultralytics
from ultralytics import YOLO
from datetime import datetime
from pathlib import Path
import shutil

def main():
# -------------------------------------------------------基本配置---------------------------------------------------------------
    project_root = Path(__file__).resolve().parent.parent
    # 路径配置
    model_cfg = project_root / f"ultralytics/cfg/models/26/yolo26-obb.yaml"
    data_yaml = project_root / f"ultralytics/cfg/datasets/DOTAv1.yaml"
 #  ##############################################################################################################################  
    # 保存路径
    project_name = project_root / f"runs/{ultralytics.__version__}"
    project_name.mkdir(parents=True, exist_ok=True)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")
    
# -------------------------------------------------------训练参数配置---------------------------------------------------------------
    cfg=dict(
        project=str(project_name),
        data=str(data_yaml),
        epochs=300,
        batch=32,
        imgsz=1024,
        
        # 其他可选配置
        workers=16,
        scale=1.0,
        mosaic=0.0,
        device="0",
        multi_scale=0.0,
        save_period=1,
        resume=False,
        amp=True,
        patience=0,
        seed=1,
        iou= 0.45,
        freeze=None,
    )
    #  ################################################################################################################################     
    # 训练 
    model = YOLO(str(model_cfg))
    model.train(**cfg)
    print(f"✅✅✅✅✅✅✅✅✅✅\033[33;41m {project_name} \033[0m✅✅✅✅✅✅✅✅✅✅")

if __name__ == "__main__":
    main()
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @colchicinewf, thank you for your detailed report and for comparing OBB training behavior across versions of Ultralytics YOLO 🚀. This is an automated response to help triage the issue, and an Ultralytics engineer will also assist you soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and answers to common questions.

Since this appears to be a 🐛 Bug Report, thank you for already including a reproducible training script and environment details 🙌. To help the team investigate the OBB convergence difference between 8.4.12 and 8.4.60 more efficiently, please also attach any available:
- complete training logs for both runs
- `results.csv` and `args.yaml`
- the exact `DOTAv1.yaml` you used after modifying paths
- confirmation that the same model config, seed, dataset split, and package versions other than `ultralytics` were used in both tests
- if possible, a smaller [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) showing the NaN behavior on a reduced subset

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify the issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the detailed comparison — please retest on the current `ultralytics` package, start from a pretrained OBB checkpoint instead of YAML-only scratch training, and try `amp=False`; if it still fails, share the full text traceback plus `ultralytics.utils.checks.collect_system_info()` output, verify your labels follow the 4-point [OBB dataset format](https://github.com/ultralytics/ultralytics/blob/main/docs/en/datasets/obb/index.md) and [OBB angle convention](https://docs.ultralytics.com/tasks/obb), delete any `labels/*.cache` files, and attach a small DOTA-split sample so we can reproduce it locally.

**Comment by Y-T-G** [comment]
@colchicinewf The mAP calculation was updated in v8.4.37, so you will see slightly lower native mAP due to that. COCOEval mAP is unaffected.

And the only other change that affects loss is https://github.com/ultralytics/ultralytics/pull/24260 which was a bug. This change was released with v8.4.39.

**Comment by Y-T-G** [comment]
@colchicinewf You can try testing this fix:

```
pip install git+https://github.com/ultralytics/ultralytics@obb_loss_fix
```

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has now been merged in [PR #24933: Fix NaN OBB loss due to sub-stride boxes](https://github.com/ultralytics/ultralytics/pull/24933), authored by @Y-T-G with contributions from @glenn-jocher.

This change may resolve the issue by improving OBB training stability when very small rotated ground-truth boxes are present. In particular, the PR:

- adds a safeguard in `ultralytics/utils/loss.py` for OBB loss preprocessing
- extracts ground-truth OBB width and height values before assignment
- replaces extremely small valid box dimensions that fall below the model stride with a stride-based fallback value
- leaves empty or invalid targets unchanged

The goal is to prevent tiny sub-stride OBB annotations from causing poor target assignment and NaN or unstable loss behavior during training.

To test the fix now, install the latest `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, it should also be included in the next official PyPI release if preferred.

If possible, please test and share whether these changes resolve the problem in your training setup. Thanks 🙏 for reporting this, and any follow-up feedback is welcome if the issue still persists.
